File: agent-app/src/main/kotlin/com/riwonace/agent/api/ChatController.kt
package com.riwonace.agent.api

import com.riwonace.agent.core.ExecutionTrace
import com.riwonace.agent.mcp.McpGateway
import com.riwonace.agent.router.Route
import com.riwonace.agent.service.AgentService
import com.riwonace.agent.service.AgentServiceV2
import jakarta.validation.Valid
import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.Size
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PostMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

/**
 * POST `/api/chat` 요청 본문.
 *
 * 보안 강화:
 * - 빈 문자열 거부 (@NotBlank)
 * - 최대 길이 제한으로 DoS 방지 (@Size)
 */
data class ChatRequest(
    @field:NotBlank(message = "질문은 공백일 수 없습니다.")
    @field:Size(min = 2, max = 2000, message = "질문은 2자 이상 2000자 이하여야 합니다.")
    val question: String,
)

/**
 * v2 응답 형식 (실행 추적 포함)
 */
data class ChatResponseV2(
    val answer: String,
    val routes: List<Route>,
    val toolCalls: List<String>,
    val contextSources: List<String>,
    val latencyMs: Long,
    /** 선택된 모델 */
    val selectedModel: String?,
    /** 클레임 커버리지 (0.0 ~ 1.0) */
    val claimCoverage: Double?,
    /** 모델 에스컬레이션 발생 여부 */
    val wasEscalated: Boolean,
    /** 실행 추적 (trace=true일 때만 포함) */
    val trace: ExecutionTraceSummary?,
)

/**
 * 실행 추적 요약 (UI 표시용)
 */
data class ExecutionTraceSummary(
    /** 총 노드 수 */
    val totalNodes: Int,
    /** 성공 노드 수 */
    val successNodes: Int,
    /** 실패 노드 수 */
    val failedNodes: Int,
    /** 스킵 노드 수 */
    val skippedNodes: Int,
    /** 노드별 실행 정보 */
    val nodes: List<NodeTraceSummary>,
    /** 계획 생성 시간 (ms) */
    val planningTimeMs: Long,
    /** 질문 의도 */
    val intent: String,
    /** 복잡도 */
    val complexity: Double,
)

data class NodeTraceSummary(
    val id: String,
    val route: String,
    val status: String,
    val durationMs: Long,
    val resultSummary: String?,
)

/**
 * 에이전트 오케스트레이션을 HTTP API로 노출하는 어댑터.
 *
 * v1: 기존 AgentService (라우팅 → 병렬 호출 → 큐레이션)
 * v2: AgentServiceV2 (프로파일링 → DAG → 최적화 → 검증)
 */
@RestController
@RequestMapping("/api")
class ChatController(
    private val agentService: AgentService,
    private val agentServiceV2: AgentServiceV2,
    private val gateway: McpGateway,
    @Value("\${agent.v2.enabled:true}")
    private val v2Enabled: Boolean,
) {
    private val log = LoggerFactory.getLogger(javaClass)

    /**
     * v1 엔드포인트: 기존 호환성 유지
     */
    @PostMapping("/chat")
    fun chat(@Valid @RequestBody request: ChatRequest): Any {
        val question = sanitizeInput(request.question)
        return if (v2Enabled) {
            val result = agentServiceV2.chat(question)
            ChatResponseV2(
                answer = result.answer,
                routes = result.routes,
                toolCalls = result.toolCalls,
                contextSources = result.contextSources,
                latencyMs = result.latencyMs,
                selectedModel = result.selectedModel,
                claimCoverage = result.claimCoverage,
                wasEscalated = result.wasEscalated,
                trace = null, // 기본 엔드포인트에서는 추적 미포함
            )
        } else {
            agentService.chat(question)
        }
    }

    /**
     * v2 엔드포인트: 실행 추적 포함
     */
    @PostMapping("/chat/v2")
    fun chatV2(
        @Valid @RequestBody request: ChatRequest,
        @RequestParam(defaultValue = "false") trace: Boolean,
    ): ChatResponseV2 {
        val question = sanitizeInput(request.question)
        val result = agentServiceV2.chat(question)

        return ChatResponseV2(
            answer = result.answer,
            routes = result.routes,
            toolCalls = result.toolCalls,
            contextSources = result.contextSources,
            latencyMs = result.latencyMs,
            selectedModel = result.selectedModel,
            claimCoverage = result.claimCoverage,
            wasEscalated = result.wasEscalated,
            trace = if (trace && result.trace != null) toTraceSummary(result.trace) else null,
        )
    }

    /** MCP 연결 상태와 K/A 분리를 진단할 수 있도록 실행 도구와 지식 리소스를 반환한다. */
    @GetMapping("/tools")
    fun tools(): Map<String, Any> =
        mapOf(
            "mcpTools" to gateway.listToolNames(),
            "mcpResources" to gateway.listResourceUris(),
        )

    /**
     * v2 상태 확인
     */
    @GetMapping("/v2/status")
    fun v2Status(): Map<String, Any> = mapOf(
        "v2Enabled" to v2Enabled,
        "features" to listOf(
            "Adaptive Model Escalation",
            "Execution DAG",
            "Evidence Budget Optimizer",
            "Answerability Gate",
            "Recovery Policy",
        ),
    )

    private fun toTraceSummary(trace: ExecutionTrace): ExecutionTraceSummary {
        return ExecutionTraceSummary(
            totalNodes = trace.plan.nodes.size,
            successNodes = trace.nodeTraces.count { it.status.name == "SUCCESS" },
            failedNodes = trace.nodeTraces.count { it.status.name == "FAILED" },
            skippedNodes = trace.nodeTraces.count { it.status.name == "SKIPPED" },
            nodes = trace.nodeTraces.map { node ->
                NodeTraceSummary(
                    id = node.nodeId,
                    route = node.route.name,
                    status = node.status.name,
                    durationMs = (node.endedAt ?: System.currentTimeMillis()) - node.startedAt,
                    resultSummary = node.resultSummary,
                )
            },
            planningTimeMs = trace.plan.planningTimeMs,
            intent = trace.plan.profile.intent.name,
            complexity = trace.plan.profile.complexity,
        )
    }

    /**
     * 입력값을 정제한다.
     * - 앞뒤 공백 제거
     * - 제어 문자 제거
     */
    private fun sanitizeInput(input: String): String {
        return input.trim()
            .replace(Regex("[\\x00-\\x1F\\x7F]"), "") // 제어 문자 제거
    }
}

File: agent-app-go/internal/api/chat_controller.go
// Package api ports com.riwonace.agent.api.ChatController from the Kotlin/Spring AI baseline.
package api

import (
	"encoding/json"
	"net/http"
	"regexp"
	"strconv"
	"strings"

	"github.com/riwonace/agent-app-go/internal/core"
	agentmcp "github.com/riwonace/agent-app-go/internal/mcp"
	"github.com/riwonace/agent-app-go/internal/service"
)

// ChatRequest mirrors data class ChatRequest: question must be 2..2000 chars, non-blank.
type ChatRequest struct {
	Question string `json:"question"`
}

// ChatResponseV2 mirrors data class ChatResponseV2.
type ChatResponseV2 struct {
	Answer         string                 `json:"answer"`
	Routes         []string               `json:"routes"`
	ToolCalls      []string               `json:"toolCalls"`
	ContextSources []string               `json:"contextSources"`
	LatencyMs      int64                  `json:"latencyMs"`
	SelectedModel  string                 `json:"selectedModel"`
	ClaimCoverage  float64                `json:"claimCoverage"`
	WasEscalated   bool                   `json:"wasEscalated"`
	Trace          *ExecutionTraceSummary `json:"trace,omitempty"`
}

// ExecutionTraceSummary mirrors data class ExecutionTraceSummary.
type ExecutionTraceSummary struct {
	TotalNodes     int                 `json:"totalNodes"`
	SuccessNodes   int                 `json:"successNodes"`
	FailedNodes    int                 `json:"failedNodes"`
	SkippedNodes   int                 `json:"skippedNodes"`
	Nodes          []NodeTraceSummary  `json:"nodes"`
	PlanningTimeMs int64               `json:"planningTimeMs"`
	Intent         string              `json:"intent"`
	Complexity     float64             `json:"complexity"`
}

// NodeTraceSummary mirrors data class NodeTraceSummary.
type NodeTraceSummary struct {
	ID            string `json:"id"`
	Route         string `json:"route"`
	Status        string `json:"status"`
	DurationMs    int64  `json:"durationMs"`
	ResultSummary string `json:"resultSummary"`
}

var controlCharsRe = regexp.MustCompile("[\x00-\x1F\x7F]")

// ChatController ports ChatController.kt 1:1 (v2-only: this port's scope targets
// AgentServiceV2, the current production path; v1/AgentService is not ported).
type ChatController struct {
	AgentServiceV2 *service.AgentServiceV2
	Gateway        *agentmcp.Gateway
}

func sanitizeInput(input string) string {
	return controlCharsRe.ReplaceAllString(strings.TrimSpace(input), "")
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.WriteHeader(status)
	json.NewEncoder(w).Encode(body)
}

func decodeChatRequest(r *http.Request) (ChatRequest, error) {
	var req ChatRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		return req, err
	}
	return req, nil
}

func validateQuestion(question string) bool {
	n := len([]rune(question))
	return strings.TrimSpace(question) != "" && n >= 2 && n <= 2000
}

// HandleChat implements POST /api/chat (v2 path only, per this port's scope).
func (c *ChatController) HandleChat(w http.ResponseWriter, r *http.Request) {
	req, err := decodeChatRequest(r)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "잘못된 요청 본문입니다."})
		return
	}
	question := sanitizeInput(req.Question)
	if !validateQuestion(question) {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "질문은 2자 이상 2000자 이하여야 합니다."})
		return
	}

	result := c.AgentServiceV2.Chat(r.Context(), question)
	writeJSON(w, http.StatusOK, toChatResponseV2(result, nil))
}

// HandleChatV2 implements POST /api/chat/v2?trace=true.
func (c *ChatController) HandleChatV2(w http.ResponseWriter, r *http.Request) {
	req, err := decodeChatRequest(r)
	if err != nil {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "잘못된 요청 본문입니다."})
		return
	}
	question := sanitizeInput(req.Question)
	if !validateQuestion(question) {
		writeJSON(w, http.StatusBadRequest, map[string]string{"error": "질문은 2자 이상 2000자 이하여야 합니다."})
		return
	}

	trace, _ := strconv.ParseBool(r.URL.Query().Get("trace"))

	result := c.AgentServiceV2.Chat(r.Context(), question)

	var traceSummary *ExecutionTraceSummary
	if trace && result.Trace != nil {
		traceSummary = toTraceSummary(result.Trace)
	}

	writeJSON(w, http.StatusOK, toChatResponseV2(result, traceSummary))
}

// HandleTools implements GET /api/tools.
func (c *ChatController) HandleTools(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	tools, err := c.Gateway.ListToolNames(ctx)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
		return
	}
	resources, err := c.Gateway.ListResourceURIs(ctx)
	if err != nil {
		writeJSON(w, http.StatusInternalServerError, map[string]string{"error": err.Error()})
		return
	}
	writeJSON(w, http.StatusOK, map[string]any{"mcpTools": tools, "mcpResources": resources})
}

// HandleV2Status implements GET /api/v2/status.
func (c *ChatController) HandleV2Status(w http.ResponseWriter, r *http.Request) {
	writeJSON(w, http.StatusOK, map[string]any{
		"v2Enabled": true,
		"features": []string{
			"Adaptive Model Escalation",
			"Execution DAG",
			"Evidence Budget Optimizer",
			"Answerability Gate",
			"Recovery Policy",
		},
	})
}

func toChatResponseV2(result service.AgentAnswerV2, trace *ExecutionTraceSummary) ChatResponseV2 {
	routes := make([]string, len(result.Routes))
	for i, r := range result.Routes {
		routes[i] = string(r)
	}
	return ChatResponseV2{
		Answer:         result.Answer,
		Routes:         routes,
		ToolCalls:      result.ToolCalls,
		ContextSources: result.ContextSources,
		LatencyMs:      result.LatencyMs,
		SelectedModel:  result.SelectedModel,
		ClaimCoverage:  result.ClaimCoverage,
		WasEscalated:   result.WasEscalated,
		Trace:          trace,
	}
}

func toTraceSummary(trace *core.ExecutionTrace) *ExecutionTraceSummary {
	var successNodes, failedNodes, skippedNodes int
	nodes := make([]NodeTraceSummary, len(trace.NodeTraces))
	for i, nt := range trace.NodeTraces {
		switch nt.Status {
		case core.NodeSuccess:
			successNodes++
		case core.NodeFailed:
			failedNodes++
		case core.NodeSkipped:
			skippedNodes++
		}
		endedAt := nt.EndedAt
		duration := endedAt - nt.StartedAt
		nodes[i] = NodeTraceSummary{
			ID:            nt.NodeID,
			Route:         string(nt.Route),
			Status:        string(nt.Status),
			DurationMs:    duration,
			ResultSummary: nt.ResultSummary,
		}
	}

	return &ExecutionTraceSummary{
		TotalNodes:     len(trace.Plan.Nodes),
		SuccessNodes:   successNodes,
		FailedNodes:    failedNodes,
		SkippedNodes:   skippedNodes,
		Nodes:          nodes,
		PlanningTimeMs: trace.Plan.PlanningTimeMs,
		Intent:         string(trace.Plan.Profile.Intent),
		Complexity:     trace.Plan.Profile.Complexity,
	}
}
